File: app/src/main/res/layout/logIn/LogInFragment.java
package com.earl.javachat.ui.logIn;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.earl.javachat.JavaChatApp;
import com.earl.javachat.R;
import com.earl.javachat.core.Keys;
import com.earl.javachat.core.OperationResultListener;
import com.earl.javachat.core.PossibleServerErrors;
import com.earl.javachat.core.SharedPreferenceManager;
import com.earl.javachat.data.restModels.LoginDto;
import com.earl.javachat.databinding.FragmentLoginBinding;
import com.earl.javachat.ui.NavigationContract;

import javax.inject.Inject;

public class LogInFragment extends Fragment implements OperationResultListener {

    FragmentLoginBinding binding;
    NavigationContract navigator;
    SharedPreferenceManager preferenceManager;
    @Inject
    LogInFormValidation validation;
    @Inject
    LogInPresenter presenter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((JavaChatApp) requireContext().getApplicationContext()).appComponent.injectLogInFragment(this);
        navigator = ((NavigationContract) requireActivity());
        preferenceManager = new SharedPreferenceManager(requireContext());
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentLoginBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.logInButton.setOnClickListener(v -> logIn());
        binding.signUpButton.setOnClickListener(v -> navigator.register());
    }

    private Boolean isValidate() {
        return validation.validate(
                binding.logInEmail,
                binding.logInPassword
        );
    }

    private void logIn() {
        if (isValidate()) {
            navigator.showProgressBar();

            LoginDto user = new LoginDto(
                    binding.logInEmail.getText().toString().trim(),
                    binding.logInPassword.getText().toString().trim()
            );
            presenter.logIn(user, this);
/*            preferenceManager.putString(Keys.KEY_TOKEN, token);
            Log.d("tag", "logIn: token -> " + token);*/
        }
    }

    @Override
    public <T> void success(T success) {
        navigator.hideProgressBar();
        preferenceManager.putBoolean(Keys.KEY_IS_SIGNED_UP, true);
        preferenceManager.putString(Keys.KEY_TOKEN, success.toString());
        Toast.makeText(requireContext(), success.toString(), Toast.LENGTH_SHORT).show();
        navigator.chat();
    }

    @Override
    public void fail(Exception exception) {
        navigator.hideProgressBar();
        if (exception.toString().equals(PossibleServerErrors.INVALID_EMAIL_OR_PASSWORD)) {
            Toast.makeText(requireContext(), R.string.invalid_pass_err, Toast.LENGTH_SHORT).show();
        } else if(exception.toString().equals(PossibleServerErrors.NO_SUCH_USER)) {
            Toast.makeText(requireContext(), R.string.no_such_user, Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(requireContext(), R.string.unknown_error, Toast.LENGTH_SHORT).show();
            Log.d("tag", "fail: " + exception);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
    }
}

File: app/src/main/java/com/earl/ktorchatapp/ui/login/LoginFragment.kt
package com.earl.ktorchatapp.ui.login

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import com.earl.ktorchatapp.KtorChatApp
import com.earl.ktorchatapp.core.BaseFragment
import com.earl.ktorchatapp.core.Keys
import com.earl.ktorchatapp.core.OperationResultListener
import com.earl.ktorchatapp.core.SharedPreferenceManager
import com.earl.ktorchatapp.databinding.FragmentLoginBinding
import com.earl.ktorchatapp.ui.NavigationContract
import com.earl.ktorchatapp.ui.ViewModelFactory
import com.earl.ktorchatapp.ui.models.UiLoginDto
import javax.inject.Inject

class LoginFragment @Inject constructor() : BaseFragment<FragmentLoginBinding>(), OperationResultListener {

    @Inject
    lateinit var validation: LoginFormValidation
    @Inject
    lateinit var viewModel: LoginFragmentViewModel
    @Inject
    lateinit var viewModelFactory: ViewModelFactory
    private lateinit var navigator: NavigationContract
    private lateinit var preferenceManager: SharedPreferenceManager

    override fun viewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ) = FragmentLoginBinding.inflate(inflater, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (requireContext().applicationContext as KtorChatApp).appComponent.injectLoginFragment(this)
        viewModel = ViewModelProvider(this, viewModelFactory)[LoginFragmentViewModel::class.java]
        navigator = requireActivity() as NavigationContract
        preferenceManager = SharedPreferenceManager(requireContext())
        binding.logInButton.setOnClickListener { logIn() }
        binding.signUpButton.setOnClickListener { navigator.register() }
    }

    private fun isValidate(): Boolean {
        return validation.validate(
            binding.logInEmail,
            binding.logInPassword
        )
    }

    private fun logIn() {
        if (isValidate()) {
            navigator.showProgressBar()
            val user = UiLoginDto.Base(
                binding.logInEmail.text.toString().trim(),
                binding.logInPassword.text.toString().trim()
            )
            viewModel.login(user, this)
        }
    }

    override fun <T> success(success: T) {
        navigator.hideProgressBar()
        preferenceManager.putBoolean(Keys.KEY_IS_SIGNED_UP, true)
        preferenceManager.putString(Keys.KEY_TOKEN, success.toString())
//        Toast.makeText(requireContext(), success.toString(), Toast.LENGTH_SHORT).show()
        navigator.chat()
    }

    override fun fail(exception: Exception?) {
        navigator.hideProgressBar()
//        Toast.makeText(requireContext(), "Error $exception", Toast.LENGTH_SHORT).show()
        Log.d("tag", "fail: error $exception")
    }

    companion object {

        fun newInstance() = LoginFragment()
    }
}